docs/source/tutorial/introduction.rst

```diff
@@ -0,0 +1,438 @@
+them. In retworkx the :class:`~retworkx.PyDiGraph` class is used to create
+directed graphs. For example:
+
+.. jupyter-execute::
+
+    import retworkx as rx
+    from retworkx.visualization import mpl_draw
+
+    path_graph = rx.generators.directed_path_graph(5)
+    mpl_draw(path_graph)
+
+In this example we created a 5 node directed path graph. This shows the
+directionality of the edges in the graph visualization with the arrow head
+pointing to the target node.
+
+Multigraphs
+===========
+
+By default all graphs in retworkx are multigraphs. This means that each
+graph object can contain parallel edges between nodes. However, you can set
+the ``multigraph`` argument to ``False`` on the :class:`~retworkx.PyGraph` and
+:class:`~retworkx.PyDiGraph` constructors when creating a new graph object to
+prevent parallel edges from being introduced. When ``multigraph`` is set to ``False``
+any method call made that would add a parallel edge will instead update the
+existing edge’s weight/data payload. For example:
+
+.. jupyter-execute::
+
+    graph = rx.PyGraph(multigraph=False)
+    graph.add_nodes_from(range(3))
+    graph.add_edges_from([(0, 1, 'A'), (0, 1, 'B'), (1, 2, 'C')])
+    mpl_draw(graph, with_labels=True, edge_labels=str)
+
+In this example, our attempt to add a parallel edge between nodes ``0``
+and ``1`` will instead result in the existing edge's data payload being updated from
+``'A'`` to ``'B'``.
+
+Graph Generators and operations
+===============================
+
+In addition to constructing graphs one node and edge at a time, you can also
+create graphs using the :ref:`generator_funcs`, :ref:`random_generators`,
+and :ref:`graph-ops` to quickly generate graphs and/or apply different
+operations on the graph. For example:
+
+.. jupyter-execute::
+
+    lolipop_graph = rx.generators.lollipop_graph(4, 3)
+    mesh_graph = rx.generators.mesh_graph(4)
+    combined_graph = rx.cartesian_product(lolipop_graph, mesh_graph)[0]
+    mpl_draw(combined_graph)
+
+Additionally there are alternate constructors such as
+:meth:`~retworkx.read_edge_list` or :meth:`~retworkx.from_adjacency_matrix`
+for building graphs from files or other inputs. For example:
+
+.. jupyter-execute::
+
+    import tempfile
+
+    with tempfile.NamedTemporaryFile('wt') as fd:
+        path = fd.name
+        fd.write('0 1\n')
+        fd.write('0 2\n')
+        fd.write('0 3\n')
+        fd.write('1 2\n')
+        fd.write('2 3\n')
+        fd.flush()
+        graph = rx.PyGraph.read_edge_list(path)
+    mpl_draw(graph)
+
+Analyzing graphs
+================
+
+The structure of a graph :math:`G` can be analyzed using the available graph algorithm
+functions. For example:
+
+.. jupyter-execute::
+
+    G = rx.PyGraph()
+    G.extend_from_edge_list([(0, 1), (0, 2)])
+    new_node = G.add_node("spam")
+    print(rx.connected_components(G))
+    degrees = []
+    for node in G.node_indices():
+        degrees.append(G.degree(node))
```

Comment:
It's probably better to store the degrees in a dict in case graph has holes in the node indices.

```suggestion
    degrees = {}
    for node in G.node_indices():
        degrees[node] = G.degree(node)
```